### PR TITLE
Introduce add_fields to add_kubernetes_metadata processor

### PR DESCRIPTION
**Pull request by kvalliyurnatt**
## Proposed commit message

This PR adds the ability to add custom fields to the events as part of the add_kuberentes_metadata processor. This reuses the add_fields processor that already exists and simply calls the add_fields processor from the add_kubernetes_metadata processor. 

The motivation behind this change to help scenarios where with packetbeat we want to see if the packet was recorded at the source or destination. By adding a custom field to denote where the packet was recorded based on the ip_port indexer in add_kubernetes_metadata processor we are able to then deduplicate some of the packets that we see on packetbeat based on the packet being recorded on the source or destination.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

This can be tested locally by using a config file like below 
```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: packetbeat-dynamic-config
  namespace: kube-system
  labels:
    k8s-app: packetbeat-dynamic
    kubernetes.io/cluster-service: "true"
data:
  packetbeat.yml: |-
    logging.level: info
    
    setup.dashboards.enabled: true
    setup.template.enabled: true
    setup.template.settings:
      index.number_of_shards: 2
    packetbeat.interfaces.device: any
    packetbeat.interfaces.type: af_packet
    packetbeat.protocols:
    - type: dns
      ports: [53]
      include_authorities: true
      include_additionals: true
    - type: http
      ports: [80, 8000, 8080, 9200]
    - type: tls
      ports: [8443]  
    - type: mysql
      ports: [3306]
    - type: redis
      ports: [6379]
    packetbeat.flows:
      enabled: true
      timeout: 30s
      period: 10s
      kill_flow_on_timeout: true
      processors:
        - community_id:
    processors:
      - add_cloud_metadata:     
      - add_kubernetes_metadata:
          node: ${HOSTNAME}
          indexers:
          - ip_port:
          matchers:
          - field_format:
              format: '%{[destination.ip]}:%{[destination.port]}'
          add_fields:
            target: PacketRecorded
            fields:
              destination: true
    cloud.auth: elastic:${ELASTIC_PASSWORD}
    cloud.id: ${CLOUD_ID}
    output.elasticsearch:
---
```

### CODE REVIEW

File: libbeat/processors/add_kubernetes_metadata/kubernetes_test.go [hunk 1]
```diff
@@ -129,6 +130,81 @@ func TestAnnotatorWithNoKubernetesAvailable(t *testing.T) {
 	assert.Equal(t, intialEventMap, event.Fields)
 }
 
+// Test field is being added correctly based on indexer and matcher
+func TestAddField(t *testing.T) {
+	cfg := config.MustNewConfigFrom(map[string]interface{}{
```
Comment (efd6) [inline libbeat/processors/add_kubernetes_metadata/kubernetes_test.go:135]:
```suggestion
	const fieldValue = "destination"
	cfg := config.MustNewConfigFrom(map[string]interface{}{
```

File: libbeat/processors/add_kubernetes_metadata/kubernetes_test.go [hunk 2]
```diff
@@ -129,6 +130,81 @@ func TestAnnotatorWithNoKubernetesAvailable(t *testing.T) {
...
+			{
+				"field_format": map[string]interface{}{
+					"format": "%{[destination.ip]}:%{[destination.port]}",
+				},
+			},
+		},
+		"add_fields": map[string]interface{}{
+			"target": "flow",
+			"fields": map[string]interface{}{
+				"destination": "destination",
```
Comment (efd6) [inline libbeat/processors/add_kubernetes_metadata/kubernetes_test.go:146]:
```suggestion
				"destination": fieldValue,
```

File: libbeat/processors/add_kubernetes_metadata/kubernetes_test.go [hunk 3]
```diff
@@ -129,6 +130,81 @@ func TestAnnotatorWithNoKubernetesAvailable(t *testing.T) {
...
+	expectedEventMap := mapstr.M{
+		"testMap": "test",
+		"destination": mapstr.M{
+			"port":    53,
+			"bytes":   408,
+			"ip":      "192.16.2.13",
+			"packets": 2,
+		},
+		"flow": mapstr.M{
+			"destination": "destination",
```
Comment (efd6) [inline libbeat/processors/add_kubernetes_metadata/kubernetes_test.go:196]:
```suggestion
			"destination": fieldValue,
```

File: libbeat/processors/add_kubernetes_metadata/kubernetes_test.go [hunk 4]
```diff
@@ -129,6 +130,81 @@ func TestAnnotatorWithNoKubernetesAvailable(t *testing.T) {
...
+		log:   logp.NewLogger(selector).With("libbeat.processor", "add_kubernetes_metadata_test"),
+		cache: annotatorCache,
+		matchers: &Matchers{
+			matchers: []Matcher{matcher},
+		},
+		kubernetesAvailable: true,
+		addFieldProcessor:   addFieldProcessor,
+	}
+
+	intialEventMap := mapstr.M{
```
Comment (efd6) [inline libbeat/processors/add_kubernetes_metadata/kubernetes_test.go:177]: Add a test case that does not match to show that it is left?

File: libbeat/processors/add_kubernetes_metadata/kubernetes.go [hunk 5]
```diff
@@ -240,6 +242,16 @@ func (k *kubernetesAnnotator) init(config kubeAnnotatorConfig, cfg *config.C) {
...
 		metaGen := metadata.GetPodMetaGen(cfg, watcher, nodeWatcher, namespaceWatcher, replicaSetWatcher, jobWatcher, metaConf)
 
+		if config.AddFields != nil {
+			k.addFieldProcessor, err = actions.CreateAddFields(config.AddFields)
+			if err != nil {
+				k.log.Errorf("Error Initializing add field processor as part of add_kubernetes_metadata processor due to error %+v", err)
+			}
+		} else {
+			//TODO: for debugging, remove or change to Debug
+			k.log.Infof("No config for addFields %+v", *cfg)
```
[the remark this answers is not in the document]
Comment (andrewkroh) [reply libbeat/processors/add_kubernetes_metadata/kubernetes.go:252]: I would remove it. My 2 🪙s.

**Review by andrewkroh — COMMENTED** [review]
Have you considered using a conditional expression instead of modifying the code? Like check if kubernetes metadata was added as a condition on `add_fields`.

```yaml
processors:
- add_kubernetes_metadata: ~
- add_fields:
    when.has_fields: [kubernetes.pod.id]
    target: foo
    fields:
      baz: bar
```

**Comment by gizas** [comment]
I am not generally a big fan for complicating the configuration for our users. I agree with the comment of https://github.com/elastic/beats/pull/37710#pullrequestreview-1840193554 .

Can all this work be done with [templates](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-autodiscover.html) or with [appenders](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-autodiscover-advanced.html) ?

**Comment by mlunadia** [comment]
doubling down on keeping the configuration as simple as possible over exposing edge cases to all users when these can be solved through other means.

**Comment by kvalliyurnatt** [comment]
@andrewkroh @gizas @mlunadia thanks for taking a look! I agree that if we can make this use case work without needing any code changes that would be ideal. However I don't see any of the current options working for us. Please correct me if I am wrong.

What we want is, in our Kubernetes environment, we want to mark the packets/flow to understand if that was recorded at the source or destination. To be able to do this we thought we could use the ip_port indexer in the add_kubernetes_metadata to say if source.ip == pod.ip then add a field to indicate that this was recorded at the source, similarly if destination.ip == pod.ip then add a field to indicate that this was recorded at the destination.
This is the config we were thinking of
```
- add_kubernetes_metadata:
          node: ${HOSTNAME}
          indexers:
          - ip_port:
          matchers:
          - field_format:
              format: '%{[destination.ip]}:%{[destination.port]}'
          add_fields:
            target: flow
            fields:
              packetRecordedAt: destination
      - add_kubernetes_metadata:
          node: ${HOSTNAME}
          indexers:
          - ip_port:
          matchers:
          - field_format:
              format: '%{[source.ip]}:%{[source.port]}'
          add_fields:
            target: flow
            fields:
              packetRecordedAt: source
```

If there are other ways to achieve this without this code change, we are happy to use that. Please let us know if some existing mechanisms could help us achieve this. Thanks!

**Comment by andrewkroh** [comment]
> Please let us know if some existing mechanisms could help us achieve this.

Based on what you describe I think a two separate add_fields processors when appropriate conditions could accomplish this. There are probably additional optimizations using `if/then` that could be implemented to avoid running the second add_kubernetes_metadata/add_fields block if the first one succeeded.

```yaml
- add_kubernetes_metadata: ~ # For destination
- add_fields:
    when.has_fields: [kubernetes.pod.id]
    target: flow
    fields:
      packetRecordedAt: destination

- add_kubernetes_metadata: ~ # For source
- add_fields:
    when.and:
      - not.has_fields: [flow.packetRecordedAt]
      - has_fields: [kubernetes.pod.id]
    target: flow
    fields:
      packetRecordedAt: source

```

**Comment by kvalliyurnatt** [comment]
> ```yaml
> when.has_fields: [kubernetes.pod.id]
>     target: flow
>     fields:
>       packetRecordedAt: destination
> ```

Thanks @andrewkroh this does work like we expected, I think this PR can be closed.

**Pull request closed without merging**